app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/LectureListTransformer.kt

```diff
@@ -0,0 +1,105 @@
+     */
+    fun legacyTransformLectureList(dayIndex: Int, lectures: List<Lecture>): LegacyLectureData {
+        val dayRoomData = transformLectureList(dayIndex, lectures)
+
+        // Set Lecture.roomIndex to match the index in DayRoomData.roomDataList
+        dayRoomData.roomDataList.forEachIndexed { index, roomData ->
+            roomData.lectures.forEach { it.roomIndex = index }
+        }
+
+        // Map room title to room index
+        val roomsMap = dayRoomData.roomDataList
+                .mapIndexed { index, roomData -> roomData.roomName to index }
+                .toMap()
+
+        // Map column index to room index
+        // This is simple identity map (0 -> 0, 1 -> 1, etc.)
+        val roomList = SparseArrayCompat<Int>()
+        for (index in 0..dayRoomData.roomDataList.size) {
+            roomList.put(index, index)
+        }
+
+        return LegacyLectureData(
+                lectureListDay = dayIndex,
+                lectureList = lectures,
+                roomsMap = roomsMap,
+                roomList = roomList,
+                roomCount = dayRoomData.roomDataList.size,
+                dayRoomData = dayRoomData
+        )
+    }
+
+    /**
+     * Transforms the given [lectures] for the given [dayIndex] into a [DayRoomData] object.
+     *
+     * Apart from the [dayIndex] it contains a list of room names and their associated lectures
+     * (sorted by [Lecture.dateUTC]). Rooms names are added in a defined order: room names of
+     * prioritized rooms first, then all other room names in the order defined by the given lecture.
+     * After adding room names the original order [Lecture.roomIndex] is no longer of interest.
+     */
+    fun transformLectureList(dayIndex: Int, lectures: List<Lecture>): DayRoomData {
+        // Pre-populate the map with prioritized rooms
+        val roomMap = prioritizedRoomProvider.prioritizedRooms
+                .map { it to mutableListOf<Lecture>() }
+                .toMap()
+                .toMutableMap()
+
+        val sortedLectures = lectures.sortedBy { it.roomIndex }
+        for (lecture in sortedLectures) {
+            val lecturesInRoom = roomMap.getOrPut(lecture.room) { mutableListOf() }
+            lecturesInRoom.add(lecture)
+        }
+
+        val roomDataList = roomMap.mapNotNull { (roomName, lectures) ->
+            if (lectures.isEmpty()) {
+                // Drop prioritized rooms without lectures
+                null
+            } else {
+                RoomData(
+                        roomName = roomName,
+                        lectures = lectures.sortedBy { it.dateUTC }.toList()
+                )
+            }
+        }
+
+        return DayRoomData(dayIndex, roomDataList)
+    }
+
+}
+
+interface PrioritizedRoomProvider {
+    val prioritizedRooms: List<String>
+}
+
+data class RoomData(
+        val roomName: String,
+        val lectures: List<Lecture>
+)
+
+data class DayRoomData(
```

Comment:
I would name this `ScheduleData` as it contains multiple rooms.

Further I would move this and `RoomData` into the `models` package as separate files. `LegacyLectureData` could stay here or could also be moved.